lib/workers/repository/process/vulnerabilities.ts

```diff
@@ -91,51 +97,372 @@ export class Vulnerabilities {
+      for (const event of this.sortEvents(range.events, versioningApi)) {
+        if (
+          is.nonEmptyString(event.introduced) &&
+          (event.introduced === '0' ||
+            (versioningApi.isVersion(event.introduced) &&
+              (versioningApi.equals(depVersion, event.introduced) ||
+                versioningApi.isGreaterThan(depVersion, event.introduced))))
+        ) {
+          vulnerable = true;
+        } else if (
+          is.nonEmptyString(event.fixed) &&
+          versioningApi.isVersion(event.fixed) &&
+          (versioningApi.equals(depVersion, event.fixed) ||
+            versioningApi.isGreaterThan(depVersion, event.fixed))
+        ) {
+          vulnerable = false;
+        } else if (
+          is.nonEmptyString(event.last_affected) &&
+          versioningApi.isVersion(event.last_affected) &&
+          versioningApi.isGreaterThan(depVersion, event.last_affected)
+        ) {
+          vulnerable = false;
+        }
+      }
+
+      if (vulnerable) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  // https://ossf.github.io/osv-schema/#evaluation
+  private isPackageVulnerable(
+    ecosystem: Ecosystem,
+    packageName: string,
+    depVersion: string,
+    affected: Osv.Affected,
+    versioningApi: VersioningApi
+  ): boolean {
+    if (this.isPackageAffected(ecosystem, packageName, affected)) {
+      if (
+        this.includedInVersions(depVersion, affected) ||
+        this.includedInRanges(depVersion, affected, versioningApi)
+      ) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private getFixedVersion(
+    ecosystem: Ecosystem,
+    depVersion: string,
+    affected: Osv.Affected,
+    versioningApi: VersioningApi
+  ): string | null {
+    const fixedVersions: string[] = [];
+    const lastAffectedVersions: string[] = [];
+
+    for (const range of affected.ranges ?? []) {
+      if (range.type === 'GIT') {
+        continue;
+      }
+
+      for (const event of range.events) {
+        if (is.nonEmptyString(event.fixed)) {
+          fixedVersions.push(event.fixed);
+        } else if (is.nonEmptyString(event.last_affected)) {
+          lastAffectedVersions.push(event.last_affected);
+        }
+      }
+    }
+
+    fixedVersions.sort((a, b) => versioningApi.sortVersions(a, b));
+    const fixedVersion = fixedVersions.find(
+      (version) =>
+        versioningApi.isVersion(version) &&
+        versioningApi.isGreaterThan(version, depVersion)
+    );
+    if (fixedVersion) {
+      return ecosystem === 'PyPI' ? `==${fixedVersion}` : fixedVersion;
+    }
+
+    lastAffectedVersions.sort((a, b) => versioningApi.sortVersions(a, b));
+    const lastAffected = lastAffectedVersions.find(
+      (version) =>
+        versioningApi.isVersion(version) &&
+        (versioningApi.equals(version, depVersion) ||
+          versioningApi.isGreaterThan(version, depVersion))
+    );
+    if (lastAffected) {
+      return `> ${lastAffected}`;
```

Comment:
I think this will only work if '>' is supported by the versioning scheme (semver for example). 

E.g., when setting a package rule like:
```
matchPackageNames: ["com.google.guava:guava"],
matchCurrentVersion: "27.1-jre"
allowedVersions: "> 29.0"
```

The logs show: 
```
Falling back to npm semver syntax for allowedVersions
"depName": "com.google.guava:guava"
```
but semver coercion / satisfy (see [here](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/process/lookup/filter.ts#L83)) would then cause the `-jre` compat suffix to be "overlooked", leading to an update PR to `guava:30.0-android`.

Hence, I think I will need to check here if `>` is supported by the scheme, e.g. using `versioningApi.isValid()`. Would be great if someone could verify whether my thoughts are correct here.

Reply:
As first step I think this is fine.    

After implementing this we should look into encapsulating the retrieval of the vulnerabilities in a function which returns a new interface. That way we could on one hand add the package rules for updates if there are fixes for them and add warnings on the dependency dashboard even if there are non currently.

Reply:
Great suggestion with showing vulnerabilities on the dashboard!

I now went ahead and changed the condition to not always blindly pick the first release. Note that this is only relevant for the subset of vulnerabilities which don't specify an exact fixed version but only the `last_affected` version. 

After further testing, my rationale is that it weighs higher that renovate proposes mergable and compatible PRs (`27.1-jre` -> `31.0-jre`), over PRs that are just one version above the no-longer-vulnerable one but faulty (`27.1-jre` -> `30.0-android`). In fact, upgrading from a vulnerable old version to the newest release shouldn't be perceived as a bug in renovate at all, so I think we should be good with keeping this for now.

Reply:
> Great suggestion with showing vulnerabilities without fixes on the dashboard!

I like this idea as well, see issue:

- #20242